plugins/auto-sizes/hooks.php

```diff
@@ -203,7 +203,23 @@ function auto_sizes_improve_image_sizes_attributes( string $content ): string {
 	}
 
 	if ( is_string( $sizes ) ) {
+		// Add the filter to update the sizes calculations to "wp_calculate_image_sizes".
+		add_filter(
+			'wp_calculate_image_sizes',
+			static function () use ( $sizes ): string {
+				return $sizes;
+			}
+		);
+
 		$processor->set_attribute( 'sizes', $sizes );
+
+		// Remove the filter to update the sizes calculations to "wp_calculate_image_sizes".
+		remove_filter(
+			'wp_calculate_image_sizes',
+			static function () use ( $sizes ): string {
+				return $sizes;
+			}
+		);
```

Comment:
I generally prefer to avoid the need to add hooks from another hook's callback, and particularly as anonymous closures, see [this section](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#closures-anonymous-functions) in the Core handbook on Closures.

> Closures should not be passed as filter or action callbacks, as removing these via remove_action() / remove_filter() is complex (at this time) (see [#46635](https://core.trac.wordpress.org/ticket/46635) for a proposal to address this). 

Additionally, I don't think this will actually do anything, because the filter is added and then removed before  `wp_calculate_image_sizes()` is ever called.

Reply:
Right, this is how this should be done:

```suggestion
		// Add the filter to update the sizes calculations to "wp_calculate_image_sizes".
		$filter = static function () use ( $sizes ): string {
			return $sizes;
		};
		add_filter(
			'wp_calculate_image_sizes',
			$filter
		);

		$processor->set_attribute( 'sizes', $sizes );

		// Remove the filter to update the sizes calculations to "wp_calculate_image_sizes".
		remove_filter(
			'wp_calculate_image_sizes',
			$filter
		);
```

A closure is an object and it needs to be the same reference when added/removed. Two identical closures are different object references.

But yeah, I also don't see how this filter is doing anything since the only code between the `add_filter()`/`remove_filter()` calls is `$processor->set_attribute( 'sizes', $sizes );` which has no side effects.

Reply:
Thanks for the feedback.

The `remove_filter` was added because of the feedback in https://github.com/WordPress/performance/pull/1382#discussion_r1687087689. As noted in https://github.com/WordPress/performance/issues/1381#issuecomment-2245637818, there is no way to get the required details in the `wp_calculate_image_sizes` filter, so we need to add the filter here to reflect the changes. Does that make sense?

Reply:
Sorry, not really. See https://github.com/WordPress/performance/pull/1382#discussion_r1690488581